OpenRA.Mods.Common/Traits/Turreted.cs

```diff
@@ -62,48 +67,95 @@ public Turreted(ActorInitializer init, TurretedInfo info)
 		{
 			this.info = info;
 			TurretFacing = GetInitialTurretFacing(init, info.InitialFacing);
+
+			if (info.RandomRealign)
+				randomFacing = init.Self.World.SharedRandom.Next(0, 256);
 		}
 
 		public void Created(Actor self)
 		{
-			attack = self.TraitOrDefault<AttackTurreted>();
 			facing = self.TraitOrDefault<IFacing>();
 			body = self.Trait<IBodyOrientation>();
 		}
 
 		public virtual void Tick(Actor self)
 		{
-			// NOTE: FaceTarget is called in AttackTurreted.CanAttack if the turret has a target.
-			if (attack != null)
+			if (!hasMoved)
 			{
-				if (!attack.IsAttacking)
+				if (info.RealignDelay != -1)
 				{
 					if (realignTick < info.RealignDelay)
 						realignTick++;
-					else if (info.RealignDelay > -1)
-						DesiredFacing = null;
-
-					MoveTurret();
+					else if (info.RandomRealign)
+					{
+						if (Face(self, randomFacing, false))
+						{
+							realignTick = 0;
+							randomFacing = self.World.SharedRandom.Next(0, 256);
+						}
+					}
+					else
+						Face(self, null, false);
 				}
+				MoveTurret();
 			}
-			else
+
+			self.World.AddFrameEndTask(w =>
+				{
+					Realign(self);
+					hasMoved = false;
+				});
+		}
+
+		void Realign(Actor self)
+		{
+			if (!hasMoved)
 			{
-				realignTick = 0;
+				if (info.RealignDelay != -1)
+				{
+					if (realignTick < info.RealignDelay)
+						realignTick++;
+					else if (info.RandomRealign)
+					{
+						if (Face(self, randomFacing, false))
+						{
+							realignTick = 0;
+							randomFacing = self.World.SharedRandom.Next(0, 256);
+						}
+					}
+					else
+						Face(self, null, false);
+				}
 				MoveTurret();
 			}
 		}
 
 		void MoveTurret()
 		{
-			var df = DesiredFacing ?? (facing != null ? facing.Facing : TurretFacing);
+			if (hasMoved)
+				return;
+
+			hasMoved = true;
+			var df = desiredFacing ?? (facing != null ? facing.Facing : TurretFacing);
 			TurretFacing = Util.TickFacing(TurretFacing, df, info.ROT);
 		}
 
-		public bool FaceTarget(Actor self, Target target)
+		public bool Face(Actor self, int? facing, bool realign = true)
 		{
-			DesiredFacing = Util.GetFacing(target.CenterPosition - self.CenterPosition, TurretFacing);
+			realignTick = realign ? 0 : realignTick;
+			desiredFacing = facing;
+
 			MoveTurret();
-			return TurretFacing == DesiredFacing.Value;
+
+			if (facing == null)
+				return TurretFacing == (facing != null ? this.facing.Facing : TurretFacing);
```

Comment:
The `facing != null` check seems redundant to me, as you are already checking `facing == null` above.